### PR TITLE
Fixes #17263 - move jed (i18n) to webpack

### PR DESCRIPTION
**Pull request by amirfefer**

### CODE REVIEW

File: webpack/assets/javascripts/bundle.js
Comment (sharvit) [inline webpack/assets/javascripts/bundle.js]: Should we name it as `foreman_i18n.js`?

File: webpack/assets/javascripts/foreman_jed.js
Comment (sharvit) [inline webpack/assets/javascripts/foreman_jed.js]: Does it make sense to add a new function to `foreman_tools` call `deprecateObjectProperty`?

```js
export const deprecateObjectProperty = (obj, oldProp, newProp, version = '1.19') =>
  Object.defineProperty(obj, oldProp, {
    get: () => {
      deprecate(oldProp, newProp, version);
      return obj[oldProp];
    },
  });

deprecateObjectProperty(window, 'i18n', 'tfm.i18n');
deprecateObjectProperty(window, 'Jed', 'tfm.i18n');
```

**Review by sharvit — CHANGES_REQUESTED** [review]
Thanks @amirfefer! love seeing staff moving into the webpack stack.

File: webpack/assets/javascripts/foreman_i18n.js
Comment (tstrachota) [inline webpack/assets/javascripts/foreman_i18n.js]: This one should imho be exported as well.

File: webpack/assets/javascripts/foreman_i18n.js
Comment (tstrachota) [inline webpack/assets/javascripts/foreman_i18n.js]: If we want to keep the wrapping blank spaces, they should be part of the prefix/suffix so that we don't pollute the translated strings when `I18N_MARK` is turned off.

File: webpack/assets/javascripts/foreman_i18n.js
Comment (tstrachota) [inline webpack/assets/javascripts/foreman_i18n.js]: Could we export Jed as i18n.jed or something similar? I'd like to avoid exporting it as default, since there's a concurrent PR, where I'm adding some more functionality:
https://github.com/theforeman/foreman/pull/5184/files#diff-1551b3b23e104a39b3f69746fe110eae

**Review by tstrachota — CHANGES_REQUESTED** [review]
Thank you @amirfefer, it works nicely. I have only couple of comments in the code.

And then I have one special request. There's a concurrent PR, where I'm adding some more i18n functionality, which is mostly complementary, but it touches the same area so there's going to be name clashes: https://github.com/theforeman/foreman/pull/5184/
It's been there for some time and currently it's blocked by broken tests, but we know that we want to merge it one day. Therefore I wanted to ask if you could make sure that both things can be merged with minimal conflicts one day. That means it should ideally use the same files and already use the same functions for the common functionality like fetching locale codes.
The most conflicting part is probably: https://github.com/theforeman/foreman/pull/5184/files#diff-1551b3b23e104a39b3f69746fe110eae
I'll be happy to help you with this.

**Comment by ohadlevy** [comment]
test foreman

**Comment by ohadlevy** [comment]
[test foreman]

File: webpack/assets/javascripts/foreman_hostgroups.js
Comment (sharvit) [inline webpack/assets/javascripts/foreman_hostgroups.js]: When i look at this file, does it make sense to call it `translate`?
```jsx
import { translate } from './foreman_i18n';

const warningMessage = translate('Some Puppet Classes are unavailable in the selected environment');
```

Some consumers can still use it as
```jsx
import { translate as __ } from './foreman_i18n';
// or
import { translate as t_ } from './foreman_i18n';
```

**Comment by ohadlevy** [comment]
@amirfefer @tstrachota what is the status of this pr? thanks

**Comment by ohadlevy** [comment]
@amirfefer what is the status of this PR? I would love to merge it prior to 1.19 branch.

**Comment by ohadlevy** [comment]
@amirfefer bump!

**Comment by amirfefer** [comment]
There were more prioritized tasks, i'll continue work on this soon.

**Comment by amirfefer** [comment]
@tstrachota , @ohadlevy  
PR rebased and updated.

**Review by sharvit — DISMISSED** [review]
LGTM! Thanks @amirfefer :+1:

File: webpack/assets/javascripts/foreman_compute_resource.js
Comment (ohadlevy) [inline webpack/assets/javascripts/foreman_compute_resource.js:33]: can we ```import { sprintf } from jed```?

**Comment by ohadlevy** [comment]
@tstrachota ping? I'm ok going forward with this with a minor nitpick mentioned above?

**Comment by ohadlevy** [comment]
@amirfefer can you rebase please? 
@tstrachota last chance to review?

File: package.json
Comment (ohadlevy) [inline package.json:76]: @amirfefer would mind trying to create a packaging pr for jed? /cc @ekohl if you have any guidance ?

**Comment by tstrachota** [comment]
@ohadlevy thanks for pinging me. I still want to take a look at this. I hope to get to it today afternoon.

File: webpack/assets/javascripts/compute_resource/openstack.js
Comment (tstrachota) [inline webpack/assets/javascripts/compute_resource/openstack.js]: `sprintf` is used quite often. Would it make sense to export it from `I18n` directly? We could then use shorter:
```js
import { sprintf, translate as __ } from '../react_app/common/I18n';
// ...
sprintf(__('Error loading scheduler hint filters information: %s'), error));
```

File: webpack/assets/javascripts/react_app/common/I18n.js
Comment (tstrachota) [inline webpack/assets/javascripts/react_app/common/I18n.js]: Shouldn't it be `this.jed.ngettext` here? Just for consistency with `translate` above.

**Review by tstrachota — COMMENTED** [review]
Good job @amirfefer I like it!

I didn't do any testing but the code looks fine.
I added two comments for consideration but in general I'm good with the current state.

**Comment by ohadlevy** [comment]
@amirfefer this needs a rebase too.

**Review by tstrachota — CHANGES_REQUESTED** [review]
Thanks for your updates @amirfefer Code-wise it looks very good now!

There's couple of js tests failing. It looks like we're missing translations at some places. Other than that it looks good to me!

**Review by tstrachota — DISMISSED** [review]
:+1: tests ran fine for me locally.

I'm waiting for Jenkins tests to confirm that.
And we still need packaging team's ack for adding the Jed. @ekohl ?

**Comment by tstrachota** [comment]
[test katello]

**Comment by ohadlevy** [comment]
[test katello]

File: package.json
Comment (ohadlevy) [reply package.json:76]: @amirfefer bump on this point?

**Review by ohadlevy — CHANGES_REQUESTED** [review]
@amirfefer looks good, please make sure you remove all of the workarounds in existing jest tests, for example:
in webpack/assets/javascripts/foreman_hostgroups.test.js, you have a block like:
```js
  beforeEach(() => {
    window.Jed = {
      sprintf(input) {
        return input;
      },
    };
    window.__ = input => input;
```

**Comment by ohadlevy** [comment]
[test katello]

**Comment by amirfefer** [comment]
@ohadlevy  done :)

**Comment by tstrachota** [comment]
@amirfefer unfortunately this needs a rebase again. Could you please take a look?

**Comment by ohadlevy** [comment]
also, please send a packaging pr to include jed?

**Comment by tstrachota** [comment]
> also, please send a packaging pr to include jed?

Btw the process is described here @amirfefer:
https://github.com/theforeman/foreman-packaging/tree/rpm/develop#adding-npm-packages
It's just a few steps.

**Comment by ohadlevy** [comment]
added packaging pr at  https://github.com/theforeman/foreman-packaging/pull/2977

**Comment by ohadlevy** [comment]
@amirfefer I've added jed to vendor file and rebased, waiting on tests.

**Comment by ohadlevy** [comment]
@amirfefer I was unable to see the deprecation warning in the browser console, is that where I'm suppose to see it? thanks!

**Comment by ohadlevy** [comment]
[test foreman]

**Comment by ohadlevy** [comment]
open / closed to retirgger tests, from some reason jenkins tests did not fire.

**Comment by ohadlevy** [comment]
[test foreman]

**Review by ohadlevy — APPROVED** [review]

**Review by ekohl — DISMISSED** [review]

File: webpack/assets/javascripts/bundle.js
Comment (ohadlevy) [inline webpack/assets/javascripts/bundle.js]: 1. i think you need to import i18n instead of jed? (otherwise tfm.i18n.translate wont work?)
2. should this be inside the react_app or somewhere in this directory? wondering as its not related to react?

File: webpack/assets/javascripts/bundle.js
Comment (ohadlevy) [inline webpack/assets/javascripts/bundle.js]: imho should just be i18n?

File: webpack/assets/javascripts/react_app/common/I18n.js
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/common/I18n.js]: I wonder if we actually need a class constructor? (since there is only one instance?) I'm OK with keeping it if it makes our life easier.

File: webpack/assets/javascripts/react_app/common/I18n.js
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/common/I18n.js]: 1. it should be
```js
deprecateObjectProperty(window, 'Jed', 'tfm.i18n.jed');
```
2. note that you didn't add any deprecations to:
```window.__``` and ```window.n__```

File: webpack/assets/javascripts/react_app/common/I18n.js
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/common/I18n.js]: please make sure these functions are tested.

**Review by ohadlevy — CHANGES_REQUESTED** [review]

File: webpack/assets/javascripts/react_app/common/I18n.js
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/common/I18n.js]: That's in purpose, in legacy js code I guess it more naturally to use `__()` then `tfm.i18.translate()` so I kept that behavior undeprecated

File: webpack/assets/javascripts/bundle.js
Comment (amirfefer) [reply webpack/assets/javascripts/bundle.js]: about the second one, I agree that it doesn't related to react.  however plugins can import foreman files via the `foremanReact` alias, which is the `react_app` folder, right?

**Comment by amirfefer** [comment]
@ohadlevy could you have another look please?

**Comment by ohadlevy** [comment]
please have a look at the tests, i believe you didn't update PaginationHelper.js file?

File: webpack/assets/javascripts/foreman_tools.js
Comment (ohadlevy) [inline webpack/assets/javascripts/foreman_tools.js:130]: can you please add a test for this new function?

File: webpack/assets/javascripts/react_app/common/I18n.js
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/common/I18n.js:14]: i think this has  no tests (for other locales) does it make sense to add one?

**Comment by ohadlevy** [comment]
@amirfefer also note the test failures, they seems related?

File: webpack/assets/javascripts/react_app/common/I18n.js
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/common/I18n.js]: this should include sprintf? (if i understand the errors correctly this is the cause...?)

**Comment by ohadlevy** [comment]
@amirfefer please rebase - thanks

**Comment by amirfefer** [comment]
@ohadlevy rebased

**Comment by ohadlevy** [comment]
@amirfefer still missing a test for https://coveralls.io/builds/19399473/source?filename=webpack/assets/javascripts/foreman_tools.js#L130 ?

File: package.json
Comment (ohadlevy) [reply package.json:76]: seems to be resolved at https://github.com/theforeman/foreman-packaging/pull/2985

**Comment by ohadlevy** [comment]
test failures not related, imho ready for merge, bonus points if you can get the missing test for depreciation.

**Review by ekohl — DISMISSED** [review]
:+1: from a packaging perspective

**Comment by amirfefer** [comment]
@ohadlevy I've added a test for `deprecateObjectProperty`

**Review by ohadlevy — APPROVED** [review]

**Comment by ohadlevy** [comment]
since @ekohl already approved, I'll merge once tests are :green_heart:

**Comment by ohadlevy** [comment]
test failures not relevant, thanks @amirfefer

**Pull request merged**
